File: algosathi/backtest.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

import pandas as pd

from algosathi.analytics import equity_curve, summarize_by_symbol
from algosathi.broker.paper_broker import PaperBroker
from algosathi.config import RiskConfig
from algosathi.core.models import Fill
from algosathi.risk.risk_manager import RiskManager
from algosathi.simulation import simulate_candles
from algosathi.strategy.base import Strategy
from algosathi.strategy.rule_strategy import RuleStrategy


@dataclass
class BacktestResult:
    total_trades: int
    realized_pnl: float
    win_rate: float
    max_drawdown: float
    equity_curve: list[dict[str, Any]] = field(default_factory=list)
    trades: list[Fill] = field(default_factory=list)


def _max_drawdown(curve: pd.DataFrame) -> float:
    if curve.empty:
        return 0.0
    running_max = curve["cumulative_realized_pnl"].cummax()
    drawdown = curve["cumulative_realized_pnl"] - running_max
    return max(0.0, float(-drawdown.min()))


def _win_rate(trades: list[Fill]) -> float:
    """Fraction of SELL (closing) trades that realized a profit, using the same
    weighted-average-cost bookkeeping as analytics.py."""
    avg_price: dict[str, float] = {}
    position_qty: dict[str, int] = {}
    wins = 0
    sells = 0

    for trade in trades:
        symbol = trade.symbol
        qty = position_qty.get(symbol, 0)
        avg = avg_price.get(symbol, 0.0)

        if trade.side == "buy":
            new_qty = qty + trade.quantity
            avg_price[symbol] = (avg * qty + trade.price * trade.quantity) / new_qty
            position_qty[symbol] = new_qty
        else:
            sells += 1
            if (trade.price - avg) * trade.quantity > 0:
                wins += 1
            new_qty = qty - trade.quantity
            position_qty[symbol] = new_qty
            if new_qty == 0:
                avg_price[symbol] = 0.0

    return wins / sells if sells else 0.0


def run_strategy_backtest(
    strategy: Strategy,
    symbol: str,
    candles: pd.DataFrame,
    risk_config: RiskConfig,
    starting_cash: float = 100_000.0,
) -> BacktestResult:
    """Runs any Strategy through the exact same simulation loop the live bot uses
    (simulation.py), then computes performance metrics. Note: an open position at the end of
    the candle window is not marked-to-market — realized_pnl only reflects closed trades."""
    trades: list[Fill] = []
    risk_manager = RiskManager(
        order_quantity=risk_config.order_quantity,
        max_daily_loss=risk_config.max_daily_loss,
        max_open_positions=risk_config.max_open_positions,
    )
    broker = PaperBroker(starting_cash=starting_cash, trade_recorder=trades.append)

    simulate_candles(strategy, risk_manager, broker, symbol, candles)

    curve = equity_curve(trades)
    summary = summarize_by_symbol(trades)

    return BacktestResult(
        total_trades=len(trades),
        realized_pnl=float(summary["realized_pnl"].sum()) if not summary.empty else 0.0,
        win_rate=_win_rate(trades),
        max_drawdown=_max_drawdown(curve),
        equity_curve=curve.to_dict("records") if not curve.empty else [],
        trades=trades,
    )


def run_backtest(
    definition: dict[str, Any],
    symbol: str,
    candles: pd.DataFrame,
    risk_config: RiskConfig,
    starting_cash: float = 100_000.0,
) -> BacktestResult:
    """Backtests a rule-tree definition — the format the web builder produces."""
    return run_strategy_backtest(
        RuleStrategy(symbol=symbol, definition=definition),
        symbol,
        candles,
        risk_config,
        starting_cash,
    )

File: web/lib/backtest.ts
// Port of algosathi/backtest.py + algosathi/simulation.py: feeds candles through the rule
// engine one at a time, filling at the *next* candle's open (avoids look-ahead bias), then
// reuses analytics.ts's P&L math (same shape as trades from Supabase) for metrics.
import { equityCurve, summarizeBySymbol } from "./analytics";
import {
  evaluateStrategy,
  type Candle,
  type SignalType,
  type StrategyDefinition,
} from "./rule-engine";
import type { Trade } from "./supabase";

export type RiskConfig = {
  orderQuantity: number;
  maxDailyLoss: number;
  maxOpenPositions: number;
};

export type BacktestResult = {
  totalTrades: number;
  realizedPnl: number;
  winRate: number;
  maxDrawdown: number;
  equityCurve: { timestamp: string; cumulativeRealizedPnl: number }[];
  trades: Trade[];
};

/** Decides what to do given all candles up to and including the current one. */
export type Evaluator = (history: Candle[]) => SignalType;

export function runBacktest(
  definition: StrategyDefinition,
  symbol: string,
  candles: Candle[],
  risk: RiskConfig
): BacktestResult {
  return runBacktestWith((history) => evaluateStrategy(definition, history), symbol, candles, risk);
}

export function runBacktestWith(
  evaluate: Evaluator,
  symbol: string,
  candles: Candle[],
  risk: RiskConfig
): BacktestResult {
  let quantity = 0;
  let avgPrice = 0;
  let realizedPnl = 0;
  const trades: Trade[] = [];

  const record = (side: "buy" | "sell", qty: number, price: number, timestamp: string) => {
    trades.push({
      id: trades.length,
      order_id: `bt-${trades.length}`,
      symbol,
      side,
      quantity: qty,
      price,
      timestamp,
      mode: "backtest",
    });
  };

  for (let i = 0; i < candles.length - 1; i++) {
    const history = candles.slice(0, i + 1);
    const signal = evaluate(history);
    if (!signal) continue;

    const price = candles[i + 1].open;
    const timestamp = candles[i + 1].timestamp;

    if (signal === "buy") {
      if (quantity > 0) continue; // already long
      if (realizedPnl <= -Math.abs(risk.maxDailyLoss)) continue; // daily loss limit
      const openPositionCount = 0; // quantity is 0 here (already-long case returned above)
      if (openPositionCount >= risk.maxOpenPositions) continue;
      const qty = risk.orderQuantity;
      const newQty = quantity + qty;
      avgPrice = (avgPrice * quantity + price * qty) / newQty;
      quantity = newQty;
      record("buy", qty, price, timestamp);
    } else {
      if (quantity <= 0) continue; // flat, nothing to exit
      const qty = quantity;
      realizedPnl += (price - avgPrice) * qty;
      quantity = 0;
      avgPrice = 0;
      record("sell", qty, price, timestamp);
    }
  }

  const curve = equityCurve(trades);
  const summary = summarizeBySymbol(trades);

  let wins = 0;
  let sells = 0;
  {
    let qty = 0;
    let avg = 0;
    for (const t of trades) {
      if (t.side === "buy") {
        const newQty = qty + t.quantity;
        avg = (avg * qty + t.price * t.quantity) / newQty;
        qty = newQty;
      } else {
        sells++;
        if ((t.price - avg) * t.quantity > 0) wins++;
        qty -= t.quantity;
        if (qty === 0) avg = 0;
      }
    }
  }

  let maxDrawdown = 0;
  {
    let runningMax = -Infinity;
    for (const p of curve) {
      runningMax = Math.max(runningMax, p.cumulativeRealizedPnl);
      maxDrawdown = Math.max(maxDrawdown, runningMax - p.cumulativeRealizedPnl);
    }
  }

  return {
    totalTrades: trades.length,
    realizedPnl: summary.reduce((sum, s) => sum + s.realizedPnl, 0),
    winRate: sells ? wins / sells : 0,
    maxDrawdown,
    equityCurve: curve,
    trades,
  };
}
